lib/src/client/connection.dart

```diff
@@ -13,11 +14,63 @@ class Connection {
   /// The stream socket subscription.
   final StreamSubscription<List<int>> _subscription;
 
+  /// Error handler from the latest listener.
+  void Function(Object, StackTrace) _onErrorListener;
+
+  /// Implementation of done.
+  final Completer<void> _done = Completer<void>();
+
   /// Creates a [Connection] instance with the given socket.
   ///
   /// [connect()] provides a more convenient way for creating instances
   /// of this class.
-  Connection(this._socket) : _subscription = _socket.listen(null);
+  Connection(this._socket) : _subscription = _socket.listen(null) {
+    // If the out-going half of the socket closes, we mark the connection as
+    // closed for sending. If there is an error, either when listening or when
+    // sending we forward it as an error to the listener.
+    // ignore: strong_mode_implicit_dynamic_parameter
+    _socket.done.then(_done.complete, onError: _onError);
+    _subscription.onError(_onError);
+  }
+
+  /// Future that is resolved when the connection is closed.
+  ///
+  /// If the connection closes gracefully this future will be completed and
+  /// future attempts to use the connection will throw a
+  /// [RedisConnectionClosedException].
+  ///
+  /// If the connection is broken, an error occurs writing/reading to/from the
+  /// connection this future will be resolved with error. If unhandled this
+  /// will propogate to the encapsulating [Zone] where it may be handled.
+  /// If using the default root [Zone] this will cause the isolate to crash.
+  /// If not implementing custom reconnection logic it might be desirable to
+  /// simply restart the process when it crashes.
+  ///
+  /// If implementing custom reconnection logic, consumers should stop using a
+  /// connection once [done] have been resolved, as all future commands will
+  /// will throw an error. If an error occurs reading or writing all outstanding
+  /// commands will be resolved with an exception.
+  Future<void> get done => _done.future;
+
+  void _onError(Object e, [StackTrace stackTrace]) {
+    log.info('an error read/write to socket occured', e, stackTrace);
+    // Stop trying to send anything new
+    if (!_done.isCompleted) {
+      _done.completeError(e, stackTrace);
```

Comment:
`_done` is completed with error, but the the exception is still thrown if `_onErrorListener` is `null`.

Is it the expected behaviour?

Reply:
Yes, if the listener doesn't listen for errors, how can the listener propagate errors, thus, there must be an unhandled exception somewhere.

It's possible that we should clean this up further in the future. I aimed this at being something that could be a minor release, as `Client` has the same interface.

Reply:
To be fair, this probably isn't perfect. Like I said, the interaction between `Connection` and `Client`/`Monitor`/etc.. is perhaps something we should re-think. So it's not possible to hijack a connection, and things like that.

Maybe we shouldn't even have `Connection` and just accept a `Socket` and then have utilities to make it easy to work with said socket inside the `Client` object...